### ISSUE TITLE
Publisher API

### ISSUE THREAD

**Issue by tyldum** [inway/mojo-rabbitmq-client#5]
AFAICS, the routing key is derived from the [?queue query-param](https://github.com/inway/mojo-rabbitmq-client/blob/master/lib/Mojo/RabbitMQ/Client/Publisher.pm#L49).
This means you need one publisher per routing key, while my design has a very dynamic routing key that is unlikely to repeat more than a few times per week.

I would like to introduce the routing key to publish(). I can create a PR, but I was debating with myself over semantics:

->publish('body',\%headers,'routing_key') vs
->publish('body', { headers => \%headers, routing_key => 'some.routing.key' })
The second one opens up for more flexibility for other options needed.

**Comment by SPodjasek** [comment]
I'll opt for following signature:
```perl
publish($body!, \%headers?, *@params)
```

which will allow to call it like this:

```perl
->publish($body, { header => 'value' }, routing_key => 'something', 
                                        mandatory => 1, immediate => 0);
->publish($body, routing_key => 'something');
->publish($body, { header => 'value' });
->publish($body);
```

With following rules:
- if $body is hashref default content_type is set to application/json and $body is encoded to JSON string, otherwise it will be passed as-is,
- if second argument is hashref it will be merged with default headers, replacing any pre-existing values, if it's defined but not a hashref it will be treated as below,
- any other (slurpy) arguments will be treated as key/value pairs and merged with AMQP publish method default parameters

What do you think?

**Comment by tyldum** [comment]
I'm fine with that!
Do you want a PR with that or would you prefer to do it yourself?

**Comment by SPodjasek** [comment]
If you can, please write a PR - I don't have much spare time lately

**Comment by tyldum** [comment]
The signature needs more work. I like the easy approach of using URL parameters, however they are not sufficient for a publisher as it can, for instance, publish to multiple exchanges.

Hows this for an API, straight from [Net::AMQP::RabbitMQ](http://search.cpan.org/~mstemle/Net-AMQP-RabbitMQ-0.310000/RabbitMQ.pm#publish($channel,_$routing_key,_$body,_$options,_$props)):
- URL parameters for exchange and routing key are defaults (if we are to keep them)
- publish($body, $routing_key, %options, %props)

Query params could still be used for defaults, but they are just as easily added with a helper.

I'm a bit time constrained as well, but I have started working on the code.

**Comment by kraih** [comment]
@tyldum Just make it flexible.

**Comment by SPodjasek** [comment]
In a5e148fc9c24a3612c6d37cf5d34dc1845200f62 I've changed publisher code to use `?routing_key=` query parameter instead of `?queue=`

**Comment by SPodjasek** [comment]
Resolved in 5a7cf91
